File: src/23/day23.py
def dfs(start: int,
        end: int,
        prev_weight: int,
        path: list[int],
        legal_moves: dict[int, list[tuple[int, int]]]) -> int:
    if start == end:
        return prev_weight
    return max(
        (dfs(move[0], end, prev_weight + move[1], path + [move[0]], legal_moves)
        for move in legal_moves[start]
        if move[0] not in path),
        default=0)

def step_forward(index: int,
                 next_square: int,
                 next_steps: dict[int, set[int]]) -> (int, int):
    step_count = 1
    possibilities = next_steps[next_square].copy()
    if index in possibilities:
        possibilities.remove(index)
    idx = next_square
    while len(possibilities) == 1:
        next_square = next(iter(possibilities))
        possibilities = next_steps[next_square].copy()
        if idx in possibilities:
            possibilities.remove(idx)
        idx = next_square
        step_count += 1
    return idx, step_count

def part1(filename: str) -> int:
    with open(filename, 'rt') as f:
        lines = [x.strip() for x in f.readlines()]
    x_bound = len(lines[0])
    y_bound = len(lines)
    starting_index = lines[0].index('.')
    ending_index = lines[-1].index('.') + x_bound * (y_bound - 1)
    grid = dict()
    i = 0
    for line in lines:
        for c in line:
            match c:
                case '.':
                    grid[i] = 0
                case '^':
                    grid[i] = -x_bound
                case '<':
                    grid[i] = -1
                case 'v':
                    grid[i] = x_bound
                case '>':
                    grid[i] = 1
            i += 1
    next_steps = dict()
    for point, direction in grid.items():
        if direction == 0:
            new_set = set()
            if point % x_bound > 0 and grid.get(point - 1, 1) != 1:
                new_set.add(point - 1)
            if point % x_bound < x_bound - 1 and grid.get(point + 1, -1) != -1:
                new_set.add(point + 1)
            if point // x_bound > 0 and grid.get(point - x_bound, x_bound) != x_bound:
                new_set.add(point - x_bound)
            if point // x_bound < y_bound - 1 and grid.get(point + x_bound, -x_bound) != -x_bound:
                new_set.add(point + x_bound)
            next_steps[point] = new_set
        else:
            if point + direction in grid:
                next_steps[point] = {point + direction}
    edges = dict()
    searching = {(starting_index, nxt) for nxt in next_steps[starting_index]}
    while searching:
        new_searching = set()
        for index, next_square in searching:
            idx, step_count = step_forward(index, next_square, next_steps)
            edges[(index, idx)] = max(step_count, edges.get((index, idx), 0))
            if idx != ending_index:
                for nxt in next_steps[idx]:
                    if nxt != index:
                        new_searching.add((idx, nxt))
        searching = new_searching
    legal_moves = dict()
    for (start, end), weight in edges.items():
        if start in legal_moves:
            legal_moves[start].append((end, weight))
        else:
            legal_moves[start] = [(end, weight)]
    return dfs(starting_index, ending_index, 0, [starting_index], legal_moves)

def step_until_junction(index: int,
                        next_square: int,
                        junctions: set[int],
                        next_steps: dict[int, set[int]]) -> (int, int):
    step_count = 1
    possibilities = next_steps[next_square].copy()
    if index in possibilities:
        possibilities.remove(index)
    idx = next_square
    while next_square not in junctions:
        next_square = next(iter(possibilities))
        possibilities = next_steps[next_square].copy()
        if idx in possibilities:
            possibilities.remove(idx)
        idx = next_square
        step_count += 1
    return idx, step_count

def part2(filename: str) -> int:
    with open(filename, 'rt') as f:
        lines = [x.strip() for x in f.readlines()]
    x_bound = len(lines[0])
    y_bound = len(lines)
    starting_index = lines[0].index('.')
    ending_index = lines[-1].index('.') + x_bound * (y_bound - 1)
    grid = dict()
    i = 0
    for line in lines:
        for c in line:
            match c:
                case '.':
                    grid[i] = 0
                case '^':
                    grid[i] = -x_bound
                case '<':
                    grid[i] = -1
                case 'v':
                    grid[i] = x_bound
                case '>':
                    grid[i] = 1
            i += 1
    junctions = {ending_index}
    next_steps = dict()
    for point, direction in grid.items():
        count = 0
        new_set = set()
        if point % x_bound > 0:
            if grid.get(point - 1, 1) != 1:
                new_set.add(point - 1)
                count += 1
            elif grid.get(point - 1, None) is not None:
                count += 1
        if point % x_bound < x_bound - 1:
            if grid.get(point + 1, -1) != -1:
                new_set.add(point + 1)
                count += 1
            elif grid.get(point + 1, None) is not None:
                count += 1
        if point // x_bound > 0:
            if grid.get(point - x_bound, x_bound) != x_bound:
                new_set.add(point - x_bound)
                count += 1
            elif grid.get(point - x_bound, None) is not None:
                count += 1
        if point // x_bound < y_bound - 1:
            if grid.get(point + x_bound, -x_bound) != -x_bound:
                new_set.add(point + x_bound)
                count += 1
            elif grid.get(point + x_bound, None) is not None:
                count += 1
        if count > 2:
            junctions.add(point)
        next_steps[point] = new_set
    edges = dict()
    searching = {(starting_index, nxt) for nxt in next_steps[starting_index]}
    while searching:
        new_searching = set()
        for index, next_square in searching:
            idx, step_count = step_until_junction(index, next_square, junctions, next_steps)
            edges[(index, idx)] = max(step_count, edges.get((index, idx), 0))
            if idx != ending_index:
                for nxt in next_steps[idx]:
                    if nxt != index:
                        new_searching.add((idx, nxt))
        searching = new_searching
    legal_moves = dict()
    for (start, end), weight in edges.items():
        if start in legal_moves:
            legal_moves[start].append((end, weight))
        else:
            legal_moves[start] = [(end, weight)]
        if end in legal_moves:
            legal_moves[end].append((start, weight))
        else:
            legal_moves[end] = [(start, weight)]
    return dfs(starting_index, ending_index, 0, [starting_index], legal_moves)

if __name__ == '__main__':
    print(part1('day23_input.txt'))
    print(part2('day23_input.txt'))

File: src/23/day23.rs
use std::collections::{BTreeSet, HashMap};

#[derive(Clone, Copy)]
pub enum Tile {
    Empty,
    Up,
    Left,
    Down,
    Right
}

fn tile_eq(t1: &Tile, t2: &Tile) -> bool {
    match (&t1, &t2) {
        (Tile::Empty,   Tile::Empty)
        | (Tile::Up,    Tile::Up)
        | (Tile::Left,  Tile::Left)
        | (Tile::Down,  Tile::Down)
        | (Tile::Right, Tile::Right) => true,
        _ => false
    }
}

fn dfs(start: usize,
       end: usize,
       prev_weight: u64,
       traversed: &mut Vec<usize>,
       legal_moves: &HashMap<usize, Vec<(usize, u64)>>) -> u64 {
    if start == end { return prev_weight; }
    return legal_moves
        .get(&start)
        .unwrap()
        .iter()
        .fold(
            0,
            |acc, &legal_move| {
                if !traversed.contains(&legal_move.0) {
                    traversed.push(legal_move.0);
                    let result = dfs(
                        legal_move.0,
                        end,
                        prev_weight + legal_move.1,
                        traversed,
                        legal_moves
                    );
                    traversed.pop();
                    std::cmp::max(acc, result)
                } else {
                    acc
                }
            }
        );
}

fn step_forward(index: usize,
                next_square: usize,
                next_steps: &HashMap<usize, BTreeSet<usize>>) -> (usize, u64) {
    let mut next_sq = next_square;
    let mut step_count = 1u64;
    let mut possibilities: BTreeSet<usize> = next_steps
        .get(&next_square)
        .unwrap()
        .iter()
        .cloned()
        .collect();
    possibilities.remove(&index);
    let mut idx = next_sq;
    while possibilities.len() == 1 {
        next_sq = possibilities.pop_first().unwrap();
        possibilities = next_steps
            .get(&next_sq)
            .unwrap()
            .iter()
            .cloned()
            .collect();
        possibilities.remove(&idx);
        idx = next_sq;
        step_count += 1;
    }
    return (idx, step_count);
}

fn part1(filename: &str) -> u64 {
    let file_text = std::fs::read_to_string(filename).unwrap();
    let lines = file_text.trim().split('\n').map(|x| x.trim()).collect::<Vec<_>>();
    let x_bound = lines[0].len();
    let y_bound = lines.len();
    let starting_index = lines[0]
        .chars()
        .enumerate()
        .find_map(|(i, c)| if c == '.' { Some(i) } else { None })
        .unwrap();
    let ending_index = lines[lines.len() - 1]
        .chars()
        .enumerate()
        .find_map(|(i, c)| if c == '.' { Some(i) } else { None })
        .unwrap()
        + x_bound * (y_bound - 1);
    let mut grid = HashMap::<usize, Tile>::new();
    let mut i: usize = 0;
    for line in lines {
        for c in line.chars() {
            match c {
                '.' => grid.insert(i, Tile::Empty),
                '^' => grid.insert(i, Tile::Up),
                '<' => grid.insert(i, Tile::Left),
                'v' => grid.insert(i, Tile::Down),
                '>' => grid.insert(i, Tile::Right),
                _   => None
            };
            i += 1;
        }
    }
    let mut next_steps = HashMap::<usize, BTreeSet<usize>>::new();
    for (&point, &direction) in &grid {
        match direction {
            Tile::Empty => {
                let mut new_set = BTreeSet::<usize>::new();
                if point % x_bound > 0
                && !tile_eq(grid.get(&(point - 1)).unwrap_or(&Tile::Right), &Tile::Right) {
                    new_set.insert(point - 1);
                }
                if point % x_bound < x_bound - 1
                && !tile_eq(grid.get(&(point + 1)).unwrap_or(&Tile::Left), &Tile::Left) {
                    new_set.insert(point + 1);
                }
                if point / x_bound > 0
                && !tile_eq(grid.get(&(point - x_bound)).unwrap_or(&Tile::Down), &Tile::Down) {
                    new_set.insert(point - x_bound);
                }
                if point / x_bound < y_bound - 1
                && !tile_eq(grid.get(&(point + x_bound)).unwrap_or(&Tile::Up), &Tile::Up) {
                    new_set.insert(point + x_bound);
                }
                next_steps.insert(point, new_set);
            },
            Tile::Up => if grid.contains_key(&(point - x_bound)) {
                next_steps.insert(point, [point - x_bound].into_iter().collect());
            },
            Tile::Left => if grid.contains_key(&(point - 1)) {
                next_steps.insert(point, [point - 1].into_iter().collect());
            },
            Tile::Down => if grid.contains_key(&(point + x_bound)) {
                next_steps.insert(point, [point + x_bound].into_iter().collect());
            },
            Tile::Right => if grid.contains_key(&(point + 1)) {
                next_steps.insert(point, [point + 1].into_iter().collect());
            }
        }
    }
    let mut edges = HashMap::<(usize, usize), u64>::new();
    let mut searching = next_steps
        .get(&starting_index)
        .unwrap()
        .iter()
        .map(|&nxt| (starting_index, nxt))
        .collect::<BTreeSet<(usize, usize)>>();
    while searching.len() > 0 {
        let mut new_searching = BTreeSet::<(usize, usize)>::new();
        for &(index, next_square) in &searching {
            let (idx, step_count) = step_forward(index, next_square, &next_steps);
            edges.insert(
                (index, idx),
                std::cmp::max(step_count, *edges.get(&(index, idx)).unwrap_or(&0))
            );
            if idx != ending_index {
                for &nxt in next_steps.get(&idx).unwrap() {
                    if nxt != index { new_searching.insert((idx, nxt)); }
                }
            }
        }
        searching = new_searching;
    }
    let mut legal_moves = HashMap::<usize, Vec<(usize, u64)>>::new();
    for (&(start, end), &weight) in &edges {
        if legal_moves.contains_key(&start) {
            legal_moves.get_mut(&start).unwrap().push((end, weight));
        } else {
            legal_moves.insert(start, vec![(end, weight)]);
        }
    }
    let mut traversed = vec![starting_index];
    return dfs(starting_index, ending_index, 0, &mut traversed, &legal_moves);
}

fn step_until_junction(index: usize,
                       next_square: usize,
                       junctions: &BTreeSet<usize>,
                       next_steps: &HashMap<usize, BTreeSet<usize>>) -> (usize, u64) {
    let mut next_sq = next_square;
    let mut step_count = 1u64;
    let mut possibilities: BTreeSet<usize> = next_steps
        .get(&next_square)
        .unwrap()
        .iter()
        .cloned()
        .collect();
    possibilities.remove(&index);
    let mut idx = next_sq;
    while !junctions.contains(&next_sq) {
        next_sq = possibilities.pop_first().unwrap();
        possibilities = next_steps
            .get(&next_sq)
            .unwrap()
            .iter()
            .cloned()
            .collect();
        possibilities.remove(&idx);
        idx = next_sq;
        step_count += 1;
    }
    return (idx, step_count);
}

fn part2(filename: &str) -> u64 {
    let file_text = std::fs::read_to_string(filename).unwrap();
    let lines = file_text.trim().split('\n').map(|x| x.trim()).collect::<Vec<_>>();
    let x_bound = lines[0].len();
    let y_bound = lines.len();
    let starting_index = lines[0]
        .chars()
        .enumerate()
        .find_map(|(i, c)| if c == '.' { Some(i) } else { None })
        .unwrap();
    let ending_index = lines[lines.len() - 1]
        .chars()
        .enumerate()
        .find_map(|(i, c)| if c == '.' { Some(i) } else { None })
        .unwrap()
        + x_bound * (y_bound - 1);
    let mut grid = HashMap::<usize, Tile>::new();
    let mut i: usize = 0;
    for line in lines {
        for c in line.chars() {
            match c {
                '.' => grid.insert(i, Tile::Empty),
                '^' => grid.insert(i, Tile::Up),
                '<' => grid.insert(i, Tile::Left),
                'v' => grid.insert(i, Tile::Down),
                '>' => grid.insert(i, Tile::Right),
                _   => None
            };
            i += 1;
        }
    }
    let mut junctions: BTreeSet<usize> = [ending_index].into_iter().collect();
    let mut next_steps = HashMap::<usize, BTreeSet<usize>>::new();
    for (&point, &_) in &grid {
        let mut count = 0;
        let mut new_set = BTreeSet::<usize>::new();
        if point % x_bound > 0 {
            if let Some(tile) = grid.get(&(point - 1)) {
                count += 1;
                if !tile_eq(tile, &Tile::Right) { new_set.insert(point - 1); }
            }
        }
        if point % x_bound < x_bound - 1 {
            if let Some(tile) = grid.get(&(point + 1)) {
                count += 1;
                if !tile_eq(tile, &Tile::Left) { new_set.insert(point + 1); }
            }
        }
        if point / x_bound > 0 {
            if let Some(tile) = grid.get(&(point - x_bound)) {
                count += 1;
                if !tile_eq(tile, &Tile::Down) { new_set.insert(point - x_bound); }
            }
        }
        if point / x_bound < y_bound - 1 {
            if let Some(tile) = grid.get(&(point + x_bound)) {
                count += 1;
                if !tile_eq(tile, &Tile::Up) { new_set.insert(point + x_bound); }
            }
        }
        if count > 2 { junctions.insert(point); }
        next_steps.insert(point, new_set);
    }
    let mut edges = HashMap::<(usize, usize), u64>::new();
    let mut searching = next_steps
        .get(&starting_index)
        .unwrap()
        .iter()
        .map(|&nxt| (starting_index, nxt))
        .collect::<BTreeSet<(usize, usize)>>();
    while searching.len() > 0 {
        let mut new_searching = BTreeSet::<(usize, usize)>::new();
        for &(index, next_square) in &searching {
            let (idx, step_count) = step_until_junction(index, next_square, &junctions, &next_steps);
            edges.insert(
                (index, idx),
                std::cmp::max(step_count, *edges.get(&(index, idx)).unwrap_or(&0))
            );
            if idx != ending_index {
                for &nxt in next_steps.get(&idx).unwrap() {
                    if nxt != index { new_searching.insert((idx, nxt)); }
                }
            }
        }
        searching = new_searching;
    }
    let mut legal_moves = HashMap::<usize, Vec<(usize, u64)>>::new();
    for (&(start, end), &weight) in &edges {
        if legal_moves.contains_key(&start) {
            legal_moves.get_mut(&start).unwrap().push((end, weight));
        } else {
            legal_moves.insert(start, vec![(end, weight)]);
        }
        if legal_moves.contains_key(&end) {
            legal_moves.get_mut(&end).unwrap().push((start, weight));
        } else {
            legal_moves.insert(end, vec![(start, weight)]);
        }
    }
    let mut traversed = vec![starting_index];
    return dfs(starting_index, ending_index, 0, &mut traversed, &legal_moves);
}

fn main() {
    println!("{}", part1("day23_input.txt"));
    println!("{}", part2("day23_input.txt"));
}
